core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksBlockMetaStore.java

```diff
@@ -275,15 +276,17 @@ public Optional<BlockMeta> getBlock(long id) {
     try {
       meta = db().get(mBlockMetaColumn.get(), Longs.toByteArray(id));
     } catch (RocksDBException e) {
-      throw new RuntimeException(e);
+      ProcessUtils.fatalError(LOG, e, "RocksDB failed");
+      throw new RuntimeException(e); // fatalError will usually system.exit
```

Comment:
i'm not quite comfortable with throwing a fatal error here. Can rocksDB throws a transient error? 
Or can we handle get() and put() errors separately? I think a read error isn't really worth a crash.

Reply:
@tcrain wdyt

Reply:
I remember there was an issue with one of the journal threads trying to exit due to an error which would cause a deadlock of waiting because we found that with the way the code was written it was only safe to shut down the master from the main thread. I can't remember if this was calling System.exit() or was just calling close on the master though.

I think the conclusion at that time was that we didn't really have a safe way to shut down the system in a general way in case of a error so we just sort of came up with a workaround for that case and marked it as something that had to be redesigned. I think @jenoudet might have better context on this though since he fixed it.

Reply:
Yes indeed Tyler, the `System.exit()` call in `UfsJournalCheckpointThread` used to cause a deadlock when encountering an error (see Alluxio/alluxio#15699). 

In general, I do not think subcomponents of the Master Process should be able to bring down the system using directly using a `System.exit()` call. If the exception should crash the master, then the exception should be a `RuntimeException` (or more accurately an `AlluxioRuntimeException`) that is not declared (in a `throws` clause) or caught anywhere.
With this in mind, maybe the solution is to have the inode store and the block store both throw exceptions, rather than both call `System.exit()`.

Reply:
Defining that all subcomponents must throw AlluxioRuntimeException and the main thread catches that and init the shutdown sequence sounds quite promising to me. But the problem might be it's not a declared exception so we really have to design systematically where to expect and catch it. It may also be hard to quickly implement at this moment because that design hasn't taken shape yet.

Therefore I think I'll just abandon this change, and let the current logic crash the master as-is. WDYT? We will come back to this in a separate discussion. If we agree on a certain design I think implementation should be relatively straightforward.

Reply:
Logged in https://github.com/Alluxio/alluxio/issues/16381, let's move the discussion there.